docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+```
+- `Sequence code` is the 64 bit sequence number assigned by `cc-agent` for a stream.
+- `Length` is the size of the entire stream message in bytes and encoded in network
+  order.
+- `Payload` is the stream data.
+
+The `cc-agent` makes use of [`libcontainer`](https://github.com/opencontainers/runc/tree/master/libcontainer)
+to manage the lifecycle of the container. This way the `cc-agent` reuses most
+of the code used by [`runc`](https://github.com/opencontainers/runc).
+
+## Runtime
+
+`cc-runtime` is an OCI compatible container runtime and is responsible for handling
+all commands specified by
+[the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+`cc-runtime` heavily utilizes the
+[virtcontainers project](https://github.com/containers/virtcontainers), which
+provides a generic, runtime-specification agnostic, hardware-virtualized containers
+library.
+
+### Configuration
+
+The runtime uses a TOML format configuration file called `configuration.toml`. By
+default this file is installed in the `/etc/clear-containers` directory.
+Most users will not need to modify the configuration file.
+
+The file is well commented and provides a few "knobs" that can modify the
+behaviour of the runtime.
+
+The configuration file is also used to enable runtime debug output (see
+https://github.com/clearcontainers/runtime#debugging).
+
+### Significant OCI commands
+
+Here we will describe how `cc-runtime` handles the most important OCI commands.
+
+#### [`create`](https://github.com/clearcontainers/runtime/blob/master/create.go)
+
+When handling the OCI `create` command, `cc-runtime` goes through the following steps:
+
+1. Create the networking container namespace on the host, according to the container
+  OCI configuration file. We only support networking namespaces for now, but
+  will support more of them later.
+2. Run all the [prestart OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks)
+  in the host namespaces created in step 1, as described by the OCI container
+  configuration file.
+3. [Set up the container networking namespace up](#networking). This is when
+  all networking interfaces are created and setup inside the namespace, according
+  to the selected pod networking model (CNM for Docker or CNI for Kubernetes).
+4. Create and start the virtual machine running the container process. The VM
+  will run inside the host namespaces created during step 1, and its `systemd`
+  instance will spawn the `cc-agent` daemon.
+5. Register the virtual machine with `cc-proxy`.
+6. The `cc-proxy` waits for the agent to signal that it is ready and then returns
+  a token. This token uniquely identifies a process within a container inside
+  the virtual machine.
+7. Spawn the `cc-shim` process providing two arguments:
+  `cc-shim --token $(token) --uri $(uri)`
+   * The proxy URL, which can be either a UNIX or a TCP socket.
```

Comment:
as the argument name to the shim is '--uri', maybe we should say 'proxy URI' just to be consistent.

Reply:
done